src/spatial/nearest_neighbor.cpp

```diff
@@ -0,0 +1,277 @@
+        return current_added == 0;
+    }
+
+    igraph_real_t worstDist() const {
+        if (current_added < max_neighbors || current_added == 0) {
+            return  max_distance;
+        }
+        return dists[current_added-1];
+    }
+
+};
+
+template <typename Metric, igraph_integer_t Dimension>
+static igraph_error_t neighbor_helper(
+        igraph_t *graph,
+        const igraph_matrix_t *points,
+        igraph_integer_t neighbors,
+        igraph_real_t cutoff,
+        igraph_integer_t dimension,
+        igraph_bool_t directed) {
+
+    const igraph_integer_t point_count = igraph_matrix_nrow(points);
+
+    using kdTree = nanoflann::KDTreeSingleIndexAdaptor<Metric, ig_point_adaptor, Dimension>;
+
+    ig_point_adaptor adaptor(points);
+
+    kdTree tree(dimension, adaptor, nanoflann::KDTreeSingleIndexAdaptorParams(10));
+
+    tree.buildIndex();
+
+    igraph_vector_t current_point;
+    IGRAPH_VECTOR_INIT_FINALLY(&current_point, dimension);
+
+    igraph_integer_t neighbor_count = neighbors >= 0 ? neighbors : point_count;
+
+    using resultClass = GraphBuildingResultSet;
+    resultClass results(neighbor_count, cutoff);
+    std::vector<igraph_integer_t> neighbor_set(neighbor_count);
+    std::vector<igraph_real_t>    distances(neighbor_count);
+    std::vector<igraph_integer_t> edges;
+    for (igraph_integer_t i = 0; i < point_count; i++) {
+
+        std::fill(neighbor_set.begin(), neighbor_set.end(), 0);
+        std::fill(distances.begin(), distances.end(), 0);
+
+        results.init(distances.data(), neighbor_set.data(), i);
+        IGRAPH_CHECK(igraph_matrix_get_row(points, &current_point, i));
+
+        tree.findNeighbors(results, VECTOR(current_point), nanoflann::SearchParameters(0, false));
+        for (igraph_integer_t j = 0; j < results.size(); j++) {
+            edges.push_back(i);
+            edges.push_back(neighbor_set[j]);
+        }
+    }
+
+    igraph_vector_destroy(&current_point);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    if (edges.size() > 2 * IGRAPH_ECOUNT_MAX) {
+        // Overflow check
+        IGRAPH_ERROR("Too many edges.", IGRAPH_EOVERFLOW);
+    }
+
+    igraph_vector_int_t edge_view;
+    igraph_vector_int_view(&edge_view, edges.data(), edges.size());
+    IGRAPH_CHECK(igraph_create(graph, &edge_view, point_count, true));
+
+    if (! directed) {
+        IGRAPH_CHECK(igraph_to_undirected(graph, IGRAPH_TO_UNDIRECTED_COLLAPSE, NULL));
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+
+template <typename Metric>
+static igraph_error_t dimension_dispatcher(
+        igraph_t *graph,
+        const igraph_matrix_t *points,
+        igraph_integer_t neighbors,
+        igraph_real_t cutoff,
+        igraph_integer_t dimension,
+        igraph_bool_t directed) {
+
+    switch (dimension) {
+    case 0:
+        if (igraph_matrix_nrow(points) == 0) {
+            // null matrix, should not error
+            igraph_empty(graph, 0, true);
+            return IGRAPH_SUCCESS;
+        }
```

Comment:
There are some issues here:

 - Return value is not checked.
 - `directed` parameter is not respected.

We usually handle edge cases early on, in the top level function, so we don't have to think about them anymore. I'll fix this.